File: src/main/java/rars/util/GraphicsPanel.java
package rars.util;

import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;

/**
 * Class that represents the panel for visualizing and animating memory reference patterns.
 */
public final class GraphicsPanel extends JPanel {
    private final int[][] colors;
    private final int columns, rows;
    private final @NotNull Dimension size;
    
    public GraphicsPanel(final int columns, final int rows) {
        super();
        this.colors = new int[rows][columns];
        this.columns = columns;
        this.rows = rows;
        this.size = new Dimension(columns, rows);
    }

    @Override
    public Dimension getPreferredSize() {
        return this.size;
    }

    @Override
    public Dimension getMaximumSize() {
        return this.size;
    }

    @Override
    public Dimension getMinimumSize() {
        return this.size;
    }

    public void paintPixel(final int row, final int col, final int color) {
        this.colors[row][col] = color;
        this.repaint(col, this.rows - row - 1, 1, 1);
    }

    @Override
    public void paint(final @NotNull Graphics g) {
        // override default paint method to assure display updated correctly every time
        // the panel is repainted.
        for (int row = 0; row < this.rows; row++) {
            for (int col = 0; col < this.columns; col++) {
                final var color = this.colors[row][col];
                g.setColor(new Color(color));
                g.fillRect(col, this.rows - row - 1, 1, 1);
            }
        }
    }
}

File: src/main/java/rars/util/GraphicsPanel.kt
package rars.util

import java.awt.Dimension
import java.awt.Graphics
import javax.swing.JPanel

/**
 * Class that represents the panel for visualizing and animating memory reference patterns.
 */
class GraphicsPanel @JvmOverloads constructor(
    size: Dimension,
    private val grid: Grid = Grid(size.height, size.width)
) : JPanel() {
    init {
        minimumSize = size
        preferredSize = size
        maximumSize = size
    }

    override fun paint(g: Graphics) {
        for (row in 0..<grid.rows) {
            for (col in 0..<grid.columns) {
                val color = this.grid.grid[row][col]
                g.color = color
                g.fillRect(col, grid.rows - row - 1, 1, 1)
            }
        }
    }
}
